internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,558 @@
+	"github.com/rs/zerolog"
+)
+
+type Peer struct {
+	conn            net.Conn
+	addr            *net.TCPAddr
+	cfg             *config.P2PConfig
+	chainParams     *chaincfg.Params
+	checkpoints     []chaincfg.Checkpoint
+	nextCheckpoint  *chaincfg.Checkpoint
+	headersService  service.Headers
+	chainService    service.Chains
+	log             *zerolog.Logger
+	services        wire.ServiceFlag
+	protocolVersion uint32
+	nonce           uint64
+	lastBlock       int32
+	startingHeight  int32
+	timeOffset      int64
+	userAgent       string
+	verAckReceived  bool
+	synced          bool
+
+	wg       sync.WaitGroup
+	msgChan  chan wire.Message
+	quitting bool
+	quit     chan struct{}
+}
+
+func NewPeer(
+	addr string,
+	cfg *config.P2PConfig,
+	chainParams *chaincfg.Params,
+	headersService service.Headers,
+	chainService service.Chains,
+	log *zerolog.Logger,
+) (*Peer, error) {
+	port, err := strconv.Atoi(config.ActiveNetParams.DefaultPort)
+	if err != nil {
+		return nil, err
+	}
+
+	ip := net.ParseIP(addr)
+	if ip == nil {
+		return nil, errors.New("could not parse peer IP")
+	}
+
+	netAddr := &net.TCPAddr{
+		IP:   ip,
+		Port: port,
+	}
+
+	currentTipHeight := headersService.GetTipHeight()
+	nextCheckpoint := findNextHeaderCheckpoint(chainParams.Checkpoints, currentTipHeight)
+
+	peer := &Peer{
+		addr:            netAddr,
+		cfg:             cfg,
+		chainParams:     chainParams,
+		headersService:  headersService,
+		chainService:    chainService,
+		checkpoints:     chainParams.Checkpoints,
+		nextCheckpoint:  nextCheckpoint,
+		log:             log,
+		services:        wire.SFspv,
+		protocolVersion: initialProtocolVersion,
+		synced:          nextCheckpoint == nil,
+		wg:              sync.WaitGroup{},
+		msgChan:         make(chan wire.Message),
+		quitting:        false,
+		quit:            make(chan struct{}),
+	}
+	return peer, nil
+}
+
+func (p *Peer) Connect() error {
+	conn, err := net.Dial(p.addr.Network(), p.addr.String())
+	if err != nil {
+		p.log.Error().Msgf("error connecting to the peer, reason: %v", err)
+		return err
+	}
+	p.conn = conn
+	p.log.Info().Msgf("connected to peer: %s", p)
+	return nil
+}
+
+func (p *Peer) Disconnect() error {
+	p.log.Info().Msgf("disconnecting peer: %s", p)
+
+	p.quitting = true
+	close(p.quit)
+	err := p.conn.Close()
+	if err != nil {
+		return err
+	}
+
+	p.wg.Wait()
+	p.log.Info().Msgf("successfully disconnected peer %s", p)
+	return nil
+}
+
+func (p *Peer) Start() error {
+	err := p.negotiateProtocol()
+	if err != nil {
+		p.log.Error().Msgf("error negotiating protocol with peer %s, reason: %v", p, err)
+		return err
+	}
+
+	go p.writeMsgHandler()
+	go p.readMsgHandler()
+	go p.pingHandler()
+
+	err = p.startSync()
+	if err != nil {
+		p.log.Error().Msgf("error starting sync from peer %s, reason: %v", p, err)
+		return err
+	}
+	return nil
+}
+
+func (p *Peer) startSync() error {
+	p.log.Info().Msgf("requesting headers from peer %s", p)
+
+	var err error
+	if p.nextCheckpoint != nil {
```

Comment:
I'm thinking about concurrency issue here - because the `nextCheckpoint` field is read here (without locking or any other sync method) and "at the same time" it could be modified in the goroutine `readMsgHandler`.

Maybe it's not an issue but I prefer to ask 😄

Reply:
Got it 👍